File: frontend/src/api/changeImpact.ts
import client from "./client";

export interface AnalyzeChangeImpactRequest {
  fmea_id: string;
  node_id: string;
  node_type: string;
  node_name: string;
  change_type: "attribute" | "structural";
  field_name?: string;
  new_value?: string;
}

export interface AffectedNode {
  node_id: string;
  node_type: string;
  name: string;
  path: string[];
  impact_type: string;
  hop_distance: number;
  risk_change: Record<string, unknown> | null;
}

export interface ImpactSummary {
  total_affected: number;
  failure_modes_affected: number;
  controls_affected: number;
  ap_upgraded_count: number;
  max_hop_distance: number;
}

export interface ChangeImpactResult {
  affected_nodes: AffectedNode[];
  summary: ImpactSummary;
}

export interface ChangeImpactAnalysis {
  id: string;
  fmea_id: string;
  product_line_code: string;
  node_id: string;
  node_type: string;
  node_name: string;
  change_type: string;
  field_name: string | null;
  old_value: string | null;
  new_value: string | null;
  scope: string;
  status: string;
  impact_score: number;
  impact_result: ChangeImpactResult;
  created_by: string;
  created_at: string;
}

export async function analyzeChangeImpact(data: AnalyzeChangeImpactRequest): Promise<ChangeImpactAnalysis> {
  const resp = await client.post("/change-impact/analyze", data);
  return resp.data;
}

export async function listChangeImpacts(fmeaId: string): Promise<ChangeImpactAnalysis[]> {
  const resp = await client.get(`/change-impact/fmea/${fmeaId}`);
  return resp.data;
}

export async function listAllChangeImpacts(productLineCode?: string): Promise<ChangeImpactAnalysis[]> {
  const resp = await client.get("/change-impact", {
    params: productLineCode ? { product_line_code: productLineCode } : undefined,
  });
  return resp.data;
}

export async function getChangeImpact(id: string): Promise<ChangeImpactAnalysis> {
  const resp = await client.get(`/change-impact/${id}`);
  return resp.data;
}

File: backend/app/schemas/change_impact.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class AffectedNode(BaseModel):
    node_id: str
    node_type: str
    name: str
    path: list[str]
    impact_type: str  # "upstream" | "downstream" | "direct"
    hop_distance: int
    risk_change: dict | None


class ImpactSummary(BaseModel):
    total_affected: int
    failure_modes_affected: int
    controls_affected: int
    ap_upgraded_count: int
    max_hop_distance: int


class ChangeImpactResult(BaseModel):
    """Repository 返回的纯分析结果，不含评分（评分由 Service 单点计算）"""
    affected_nodes: list[AffectedNode]
    summary: ImpactSummary


class ChangeImpactAnalyzeRequest(BaseModel):
    fmea_id: uuid.UUID
    node_id: str
    node_type: str
    node_name: str
    change_type: Literal["attribute", "structural"]
    field_name: str | None = None
    new_value: str | None = None


class ChangeImpactAnalysisResponse(BaseModel):
    id: uuid.UUID
    fmea_id: uuid.UUID
    product_line_code: str
    node_id: str
    node_type: str
    node_name: str
    change_type: str
    field_name: str | None
    old_value: str | None
    new_value: str | None
    scope: str
    status: str
    impact_score: int
    impact_result: ChangeImpactResult
    created_by: uuid.UUID
    created_at: datetime

    model_config = {"from_attributes": True}
